sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/_shared/user_credential_async.py

```diff
@@ -3,87 +3,119 @@
+        self._refresh_proactively = kwargs.pop('refresh_proactively', False)
+        self._timer = None
+        self._async_mutex = Lock()
+        self._lock = Condition(self._async_mutex)
         self._some_thread_refreshing = False
+        if self._refresh_proactively:
+            self._schedule_refresh()
 
     async def get_token(self, *scopes, **kwargs):  # pylint: disable=unused-argument
         # type (*str, **Any) -> AccessToken
         """The value of the configured token.
         :rtype: ~azure.core.credentials.AccessToken
         """
-        if not self._token_refresher or not self._token_expiring():
+        if not self._token_refresher or not self._is_token_expiring_soon(self._token):
             return self._token
+        await self._update_token_and_reschedule()
+        return self._token
 
+    async def _update_token_and_reschedule(self):
         should_this_thread_refresh = False
-
         async with self._lock:
-
-            while self._token_expiring():
+            while self._is_token_expiring_soon(self._token):
                 if self._some_thread_refreshing:
-                    if self._is_currenttoken_valid():
+                    if self._is_token_valid(self._token):
                         return self._token
-
-                    await self._wait_till_inprogress_thread_finish_refreshing()
+                    await self._wait_till_lock_owner_finishes_refreshing()
                 else:
                     should_this_thread_refresh = True
                     self._some_thread_refreshing = True
                     break
 
-
         if should_this_thread_refresh:
             try:
-                newtoken = await self._token_refresher()  # pylint:disable=not-callable
-
+                new_token = await self._token_refresher()
+                if not self._is_token_valid(new_token):
+                    raise ValueError(
+                        "The token returned from the token_refresher is expired.")
                 async with self._lock:
-                    self._token = newtoken
+                    self._token = new_token
                     self._some_thread_refreshing = False
                     self._lock.notify_all()
             except:
                 async with self._lock:
                     self._some_thread_refreshing = False
                     self._lock.notify_all()
-
                 raise
-
+        if self._refresh_proactively:
+            self._schedule_refresh()
         return self._token
 
-    async def _wait_till_inprogress_thread_finish_refreshing(self):
+    def _schedule_refresh(self):
+        if self._timer is not None:
+            self._timer.cancel()
+
+        token_ttl = self._token.expires_on - get_current_utc_as_int()
+
+        if self._is_token_expiring_soon(self._token):
+            # Schedule the next refresh for when it reaches a certain percentage of the remaining lifetime.
+            timespan = token_ttl / 2
```

Comment:
I'm not sure how the `AsyncTimer` constructor behaves if you pass a decimal number, I'd suggest using `timespan = token_ttl // 2` to round it.